In [2]:
import pandas as pd
import numpy as np
import os, sys

In [4]:
datasets = './datasets/titanic'
os.listdir(datasets)

['gender_submission.csv', 'test.csv', 'train.csv']

In [5]:
train_file = os.path.join(datasets, 'train.csv')
test_file = os.path.join(datasets, 'test.csv')
gender_file = os.path.join(datasets, 'gender_submission.csv')

In [8]:
df_train = pd.read_csv(train_file, sep=',', encoding='u8')
df_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

一共891个乘客信息，每个乘客信息包含乘客ID、是否幸存、乘客仓位级别、姓名、性别、年龄、兄弟姐妹个数、直系亲属个数、船票信息、票价、客舱、登船窗口信息。

In [7]:
df_train_chunked = pd.read_csv(train_file, chunksize=1000)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [20]:
df_train.columns

Index(['PassengerId', 'Survived', '仓位级别', 'Name', 'Sex', 'Age', '兄弟姐妹个数',
       '直系亲属个数', 'Ticket', 'Fare', '客舱', '登船窗口'],
      dtype='object')

In [13]:
df_train.rename(columns={
    'SibSp' : '兄弟姐妹个数',
    'Parch' : '直系亲属个数',
    'Cabin' : '客舱',
    'Embarked' : '登船窗口',
    'Pclass' : '仓位级别',
}, inplace = True)
df_train

PassengerId  Survived  仓位级别  \
0              1         0     3   
1              2         1     1   
2              3         1     3   
3              4         1     1   
4              5         0     3   
..           ...       ...   ...   
886          887         0     2   
887          888         1     1   
888          889         0     3   
889          890         1     1   
890          891         0     3   

                                                  Name     Sex   Age  兄弟姐妹个数  \
0                              Braund, Mr. Owen Harris    male  22.0       1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0       1   
2                               Heikkinen, Miss. Laina  female  26.0       0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0       1   
4                             Allen, Mr. William Henry    male  35.0       0   
..                                                 ...     ...   ...     ...   
886                              Montvila, Rev. Juozas    male  27.0       0   
887                       Graham, Miss. Margaret Edith  female  19.0       0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN       1   
889                              Behr, Mr. Karl Howell    male  26.0       0   
890                                Dooley, Mr. Patrick    male  32.0       0   

     直系亲属个数            Ticket     Fare    客舱 登船窗口  
0         0         A/5 21171   7.2500   NaN    S  
1         0          PC 17599  71.2833   C85    C  
2         0  STON/O2. 3101282   7.9250   NaN    S  
3         0            113803  53.1000  C123    S  
4         0            373450   8.0500   NaN    S  
..      ...               ...      ...   ...  ...  
886       0            211536  13.0000   NaN    S  
887       0            112053  30.0000   B42    S  
888       2        W./C. 6607  23.4500   NaN    S  
889       0            111369  30.0000  C148    C  
890       0            370376   7.7500   NaN    Q  

[891 rows x 12 columns]

In [14]:
df_train.isnull()

PassengerId  Survived   仓位级别   Name    Sex    Age  兄弟姐妹个数  直系亲属个数  \
0          False     False  False  False  False  False   False   False   
1          False     False  False  False  False  False   False   False   
2          False     False  False  False  False  False   False   False   
3          False     False  False  False  False  False   False   False   
4          False     False  False  False  False  False   False   False   
..           ...       ...    ...    ...    ...    ...     ...     ...   
886        False     False  False  False  False  False   False   False   
887        False     False  False  False  False  False   False   False   
888        False     False  False  False  False   True   False   False   
889        False     False  False  False  False  False   False   False   
890        False     False  False  False  False  False   False   False   

     Ticket   Fare     客舱   登船窗口  
0     False  False   True  False  
1     False  False  False  False  
2     False  False   True  False  
3     False  False  False  False  
4     False  False   True  False  
..      ...    ...    ...    ...  
886   False  False   True  False  
887   False  False  False  False  
888   False  False   True  False  
889   False  False  False  False  
890   False  False   True  False  

[891 rows x 12 columns]

按票价和年龄两列进行综合排序（降序排列），

In [15]:
df_train.sort_values(by=['Fare', 'Age'], ascending=False)

PassengerId  Survived  仓位级别                                Name     Sex  \
679          680         1     1  Cardeza, Mr. Thomas Drake Martinez    male   
258          259         1     1                    Ward, Miss. Anna  female   
737          738         1     1              Lesurer, Mr. Gustave J    male   
438          439         0     1                   Fortune, Mr. Mark    male   
341          342         1     1      Fortune, Miss. Alice Elizabeth  female   
..           ...       ...   ...                                 ...     ...   
481          482         0     2    Frost, Mr. Anthony Wood "Archie"    male   
633          634         0     1       Parr, Mr. William Henry Marsh    male   
674          675         0     2          Watson, Mr. Ennis Hastings    male   
732          733         0     2                Knight, Mr. Robert J    male   
815          816         0     1                    Fry, Mr. Richard    male   

      Age  兄弟姐妹个数  直系亲属个数    Ticket      Fare           客舱 登船窗口  
679  36.0       0       1  PC 17755  512.3292  B51 B53 B55    C  
258  35.0       0       0  PC 17755  512.3292          NaN    C  
737  35.0       0       0  PC 17755  512.3292         B101    C  
438  64.0       1       4     19950  263.0000  C23 C25 C27    S  
341  24.0       3       2     19950  263.0000  C23 C25 C27    S  
..    ...     ...     ...       ...       ...          ...  ...  
481   NaN       0       0    239854    0.0000          NaN    S  
633   NaN       0       0    112052    0.0000          NaN    S  
674   NaN       0       0    239856    0.0000          NaN    S  
732   NaN       0       0    239855    0.0000          NaN    S  
815   NaN       0       0    112058    0.0000         B102    S  

[891 rows x 12 columns]

通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？

In [17]:
max(df_train['兄弟姐妹个数'] + df_train['直系亲属个数'])

10

分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据

In [18]:
df_train['Fare'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

In [19]:
df_train['直系亲属个数'].describe()

count    891.000000
mean       0.381594
std        0.806057
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        6.000000
Name: 直系亲属个数, dtype: float64

读取test.csv

In [21]:
df_test = pd.read_csv(test_file)
df_test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

添加多余的一列，然后删除

In [22]:
df_test['a'] = False
df_test.head(3)

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2          894       2         Myles, Mr. Thomas Francis    male  62.0      0   

   Parch  Ticket    Fare Cabin Embarked      a  
0      0  330911  7.8292   NaN        Q  False  
1      0  363272  7.0000   NaN        S  False  
2      0  240276  9.6875   NaN        Q  False

In [23]:
df_test.drop(columns=['a'], inplace=True)
df_test.head(3)

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2          894       2         Myles, Mr. Thomas Francis    male  62.0      0   

   Parch  Ticket    Fare Cabin Embarked  
0      0  330911  7.8292   NaN        Q  
1      0  363272  7.0000   NaN        S  
2      0  240276  9.6875   NaN        Q

将['PassengerId','Name','Age','Ticket']这几个列元素隐藏，只观察其他几个列元素

In [25]:
df_train.drop(columns=['PassengerId','Name','Age','Ticket']).head(3)

Survived  仓位级别     Sex  兄弟姐妹个数  直系亲属个数     Fare   客舱 登船窗口
0         0     3    male       1       0   7.2500  NaN    S
1         1     1  female       1       0  71.2833  C85    C
2         1     3  female       0       0   7.9250  NaN    S

 以"Age"为条件，将年龄在10岁以上和50岁以下的乘客信息显示出来，并将这个数据命名为midage

In [33]:
df_train_midage = df_train[(df_train['Age'] > 10) & (df_train['Age'] < 50)]
df_train_midage

PassengerId  Survived  仓位级别  \
0              1         0     3   
1              2         1     1   
2              3         1     3   
3              4         1     1   
4              5         0     3   
..           ...       ...   ...   
885          886         0     3   
886          887         0     2   
887          888         1     1   
889          890         1     1   
890          891         0     3   

                                                  Name     Sex   Age  兄弟姐妹个数  \
0                              Braund, Mr. Owen Harris    male  22.0       1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0       1   
2                               Heikkinen, Miss. Laina  female  26.0       0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0       1   
4                             Allen, Mr. William Henry    male  35.0       0   
..                                                 ...     ...   ...     ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0       0   
886                              Montvila, Rev. Juozas    male  27.0       0   
887                       Graham, Miss. Margaret Edith  female  19.0       0   
889                              Behr, Mr. Karl Howell    male  26.0       0   
890                                Dooley, Mr. Patrick    male  32.0       0   

     直系亲属个数            Ticket     Fare    客舱 登船窗口  
0         0         A/5 21171   7.2500   NaN    S  
1         0          PC 17599  71.2833   C85    C  
2         0  STON/O2. 3101282   7.9250   NaN    S  
3         0            113803  53.1000  C123    S  
4         0            373450   8.0500   NaN    S  
..      ...               ...      ...   ...  ...  
885       5            382652  29.1250   NaN    Q  
886       0            211536  13.0000   NaN    S  
887       0            112053  30.0000   B42    S  
889       0            111369  30.0000  C148    C  
890       0            370376   7.7500   NaN    Q  

[576 rows x 12 columns]

将midage的数据中第100行的"Pclass"和"Sex"的数据显示出来

In [35]:
df_train_midage.reset_index(drop=True, inplace=True)
df_train_midage

PassengerId  Survived  仓位级别  \
0              1         0     3   
1              2         1     1   
2              3         1     3   
3              4         1     1   
4              5         0     3   
..           ...       ...   ...   
571          886         0     3   
572          887         0     2   
573          888         1     1   
574          890         1     1   
575          891         0     3   

                                                  Name     Sex   Age  兄弟姐妹个数  \
0                              Braund, Mr. Owen Harris    male  22.0       1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0       1   
2                               Heikkinen, Miss. Laina  female  26.0       0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0       1   
4                             Allen, Mr. William Henry    male  35.0       0   
..                                                 ...     ...   ...     ...   
571               Rice, Mrs. William (Margaret Norton)  female  39.0       0   
572                              Montvila, Rev. Juozas    male  27.0       0   
573                       Graham, Miss. Margaret Edith  female  19.0       0   
574                              Behr, Mr. Karl Howell    male  26.0       0   
575                                Dooley, Mr. Patrick    male  32.0       0   

     直系亲属个数            Ticket     Fare    客舱 登船窗口  
0         0         A/5 21171   7.2500   NaN    S  
1         0          PC 17599  71.2833   C85    C  
2         0  STON/O2. 3101282   7.9250   NaN    S  
3         0            113803  53.1000  C123    S  
4         0            373450   8.0500   NaN    S  
..      ...               ...      ...   ...  ...  
571       5            382652  29.1250   NaN    Q  
572       0            211536  13.0000   NaN    S  
573       0            112053  30.0000   B42    S  
574       0            111369  30.0000  C148    C  
575       0            370376   7.7500   NaN    Q  

[576 rows x 12 columns]

In [36]:
df_train_midage.loc[[100], ['仓位级别', 'Sex']]

仓位级别   Sex
100     2  male

使用loc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [38]:
df_train_midage.loc[[100,105,108],['仓位级别','Name','Sex']]

仓位级别                               Name   Sex
100     2  Byles, Rev. Thomas Roussel Davids  male
105     3           Cribb, Mr. John Hatfield  male
108     3                    Calic, Mr. Jovo  male

使用iloc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [39]:
df_train_midage.iloc[[100,105,108],[2,3,4]]

仓位级别                               Name   Sex
100     2  Byles, Rev. Thomas Roussel Davids  male
105     3           Cribb, Mr. John Hatfield  male
108     3                    Calic, Mr. Jovo  male

In [ ]:
df_train.to_csv("df_train.csv")